### ISSUE TITLE
markdown: Custom templates and front matter

### ISSUE THREAD

**Issue by mholt** [mholt/caddy#35]
Right now, the markdown directive only lets you specify CSS and JS files. It would be really useful to be able to specify a full page template like you can with the browse middleware.

Markdown should also be expanded to support front matter that contains metadata, etc... it could be used to fill template values and render the page.

**Comment by abiosoft** [comment]
These are my thoughts so far.

Support for front matters, both sitewide and each page. Example below.
```toml
---
title = "Page Title"
js = [ "jquery.js", "script.js" ]  
css = [ "bootstrap.css", "style.css" ]

[variables]
name = "Matt"
date = "May 1, 2015"

[[meta]]
charset = "UTF-8"

[[meta]]
name = "description"
content = "Caddy Markdown Page"

[[meta]]
name = "keywords"
content = "caddy, markdown"

[[meta]]
name = "author"
content = "Matt Holt"
---
[Page markdown contents goes here]
```
* Front matter sample above specified in [TOML](https://github.com/toml-lang/toml) syntax. (debatable)
* Front matter can be embedded in a page by preceding and proceeding with `---` or in a separate file (see Caddyfile below).
* Title, multiple js and multiple css can be specified as shown above.
* Variables are defined under `[variables]` and will be accessible using standard Go template e.g. `{{name}}` within page contents. There can also be some useful predefined variables like current date.
* Multiple meta tags can be specified using array of dictionaries. Example above.

Caddyfile syntax, backward compatible and with new **front** keyword.

Syntax: `front path filepath`
path - request path relative to the site
filepath - path to the toml file

All pages on that path will be rendered using the config in the specified toml file. If there are conflicting configs (mainly variables), the one embedded in page will be preferred.
```
markdown /blog {
    ext .md .txt
    css /resources/css/blog.css
    js  /resources/js/blog.js
    front /projects /projects/projects.toml
    front / /blog.toml
}
```
Does this looks practical and feasible ? 

Thanks.

**Comment by mholt** [comment]
Sure does. A few comments:

- I like that idea of templates. I'm wondering if we can extend it to allow a full HTML template page (like the `browse` directive does). That will allow the user to build in any JS, CSS, keywords, meta tags, etc, that they want to. But if they don't want to use a template, it'll still useful to have the css and js properties so they can at least specify a stylesheet or script. So anyway, consider a `template` property that takes a filename as an argument. Does that make sense?
- I bet we could auto-detect front matter when embedded in the markdown file. I think Hugo does this and it's a great example. What's more is we may not be limited to just TOML - if it's somewhat trivial (maybe by using an interface), we could make that swappable for YAML or JSON depending on how the file starts (`---`, `+++`, `{`). (Or it might not start with front matter at all.)
-  I can see the utility for the `front` property - clever idea. But given the idea of templates, I wonder if the need to apply a universal front matter to a bunch of files kind of falls out. Which is good, if that will simplify things!

**Comment by abiosoft** [comment]
* It will support multiple types exactly like Hugo did, that was my first thought.
* I agree that specifying templates will be a better option.

I can start working on these now but I need three more clarifications.
* Should it still be possible to add css and js files from the Caddyfile since templates are now allowed and will include that already?
* I think it is more convenient for the user to register templates in the Caddyfile and use the aliases in the front matter. Or allow both aliases and filename.
* Should nested templates be supported ?

**Comment by mholt** [comment]
- That is a good question. I am not sure actually... it sure is convenient to just specify css/js files this way if you want to use the default template, but if you want to specify your own template, the css/js properties take no effect because Caddy isn't going to modify your template. Maybe css/js and template should be mutually exclusive (neither or one, but not both)?
- Sounds great. To keep things easy for users, it would be ideal if they could specify a template without a name, if they only want to use one template. Like `template mytemplate.html`.
- If it's not hard to do, yes.

:+1:

**Comment by abiosoft** [comment]
Almost done with this, remaining some more code documentation/refactor and tests. Anyone not willing to wait can build from my fork.

TOML, YAML and JSON metadata are supported in each Markdown page. 
Variables can be added to metadata and are accessible in a template using Go's template syntax `{{ .variable }}`.

TOML
```
+++
template = "blog"
[variables]
sitename = "A caddy site"
+++
```
YAML
```
---
template: blog
---
```
JSON
```
{
"template" : "blog"
}
```
templates can be added to Caddyfile via **template** syntax.

Usage
```
template path # register a default template
template name path # register a template with name
```
e.g. 
```
markdown /blog {
    ext .md .txt
    template blog blog.html
    template about about.html
    template default.html
}
```
If no template is specified (or no metadata is found), default registered template is used. If default registered template is not found, the current basic template will be used.

Finally, `css` and `js` directives are still supported if no template is specified and the basic template is used.

**Comment by mholt** [comment]
Closed with #68
